File: Client/Services/MessageService.cs
using WebApp.Models;
using WebApp.Repositories;
using WebApp.Util;

namespace WebApp.Services;

public class MessageService: IMessageService
{
    private readonly IMessageRepository _messageRepository;
    private readonly IUserRepository _userRepository;

    public MessageService( IMessageRepository messageRepository, IUserRepository userRepository)
    {
        _messageRepository = messageRepository;
        _userRepository = userRepository;
    }

    public async Task<IEnumerable<Message>> GetAllMessages(int userId)
    {
            return await _messageRepository.GetMessagesForUser(userId);
    }
    

    public async Task<ResponseMSG> SendMessage(string username, string content, string title)
    {
        
        try
        {
            var user = await _userRepository.GetUserByName(username);
            if (user == null)
            {
                return new ResponseMSG
                {
                    Status = false,
                    Message = "user not found",
                };
            }
            var message = new Message
            {
                Title = title,
                Content = content,
                UserId = user.Id,
                Read = false,
                CreatedAt = DateTime.UtcNow
            };

            _messageRepository.Add(message);

            return new ResponseMSG()
            {
                Status = true,
                Message = "Message added: " + message,
            };
        }
        catch (Exception ex)
        {
            return new ResponseMSG()
            {
                Status = false,
                Message = ex.Message,
            };
        }
    }

    public async Task<ResponseMSG> UpdateMessageStatus(int messageId, int userId, bool isRead)
    {
        try
        {
            var message = await  _messageRepository.GetById(messageId);

            if (message == null)
                return new ResponseMSG()
                {
                    Status = false,
                    Message = "Message not found",
                };

            if (message.UserId != userId)
                return new ResponseMSG()
                {
                    Status = false,
                    Message = "User not found",
                };

            message.Read = isRead;
            _messageRepository.Update(message);

            return new ResponseMSG()
            {
                Status = true,
                Message = "Message updated",
            };
        }
        catch (Exception ex)
        {
            return new ResponseMSG()
            {
                Status = false,
                Message = ex.Message,
            };
        }
    }
}

File: Client/src/app/services/message.service.ts
import { computed, Injectable, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { Observable ,firstValueFrom} from 'rxjs';
import { Notification } from '../models/notification.model';

@Injectable({ providedIn: 'root' })
export class MessageService {
  private apiUrl = 'http://localhost:5066/api';

  constructor(private http: HttpClient, private router: Router) {}
  notifications = signal<Notification[]>([]);
  notificationCount = computed(() => this.notifications().length);


  async ngOnInit() {
    await this.loadNotifications();
  }

  async loadNotifications() {
    try {
      const messages = await this.getMessages();
      this.notifications.set(messages);
    } catch (error) {
      console.error('Failed to load notifications:', error);
    }
  }



  async getMessages(): Promise<any> {
    try {
      const token = this.getToken();
      if (!token) {
        throw new Error("No authentication token found");
      }

      const response = await firstValueFrom(
        this.http.get(`${this.apiUrl}/messages`, {
          headers: {
            Authorization: `Bearer ${token}`
          }
        })
      );
      console.log("profile response: ", response);


      return response;
    } catch (error) {
      console.error("Get user error:", error);
      throw error;
    }
  }

  async sendMessage(username: string, title: string, content: string): Promise<any> {
    try {
      const token = this.getToken();
      if (!token) {
        this.router.navigate(['/login']);
        throw new Error("Authentication required");
      }

      const body = {
        Content: content,
        Title: title
      };

      const response = await firstValueFrom(
        this.http.post(`${this.apiUrl}/messages/${username}`, body, {
          headers: {
            Authorization: `Bearer ${token}`
          }
        })
      );

      return response;
    } catch (error) {
      console.error("Send message error:", error);
      throw error;
    }
  }

  async updateMessageStatus(id: number, isRead: boolean): Promise<Notification> {
    try {
      const token = this.getToken();
      if (!token) {
        this.router.navigate(['/login']);
        throw new Error("Authentication required");
      }

      const response = await firstValueFrom(
        this.http.patch<any>(
          `${this.apiUrl}/messages/${id}`,
          { IsRead: isRead },
          {
            headers: {
              Authorization: `Bearer ${token}`,
              'Content-Type': 'application/json'
            }
          }
        )
      );

      return this.mapToNotification(response);
    } catch (error) {
      console.error("Update message error:", error);
      throw error;
    }
  }

  private mapToNotification(msg: any): Notification {
    return {
      id: msg.id,
      title: msg.title,
      content: msg.content,
      read: msg.isRead,
      createdAt: new Date(msg.createdAt),
    };
  }

  getToken(): string | null {
    return localStorage.getItem('token');
  }
}
